demos/coreMQTT/mqtt_demo_keep_alive.c

```diff
@@ -0,0 +1,1172 @@
+ */
+static volatile bool xKeepAliveError = false;
+
+/**
+ * @brief The last time when a PINGREQ was sent over the network.
+ */
+static uint32_t ulPingReqSendTimeMs;
+
+/**
+ * @brief Timeout for a pending PINGRESP from the MQTT broker.
+ */
+static uint32_t ulPingRespTimeoutMs = ( mqttexampleKEEP_ALIVE_TIMEOUT_SECONDS / 4 ) * MILLISECONDS_PER_SECOND;
+
+/**
+ * @brief Static buffer used to hold an MQTT PINGREQ packet for keep-alive mechanism.
+ */
+const static MQTTFixedBuffer_t xPingReqBuffer =
+{
+    .pBuffer = ucPingReqBuffer,
+    .size    = MQTT_PACKET_PINGREQ_SIZE
+};
+
+/** @brief Static buffer used to hold MQTT messages being sent and received. */
+static MQTTFixedBuffer_t xBuffer =
+{
+    .pBuffer = ucSharedBuffer,
+    .size    = mqttexampleSHARED_BUFFER_SIZE
+};
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief This example uses MQTT APIs to send and receive MQTT packets over a
+ * TCP connection established using a transport interface implementation based
+ * on the Secure Sockets library. This example is single threaded and uses
+ * statically allocated memory. This example uses QOS0 and therefore does not
+ * implement any retransmission mechanism for Publish messages. This example
+ * runs for democonfigMQTT_MAX_DEMO_COUNT, if the connection to the broker goes
+ * down, the code tries to reconnect to the broker with an exponential backoff
+ * mechanism. This example also creates a Keep Alive timer task to handle
+ * periodically sending a PINGREQ to the broker. Because the demo application
+ * handles the MQTT Keep Alive, this example uses the MQTT_RecieveLoop API
+ * instead of the MQTT_ProcessLoop API used in the MQTT plaintext demo.
+ */
+int RunCoreMqttKeepAliveDemo( bool awsIotMqttMode,
+                              const char * pIdentifier,
+                              void * pNetworkServerInfo,
+                              void * pNetworkCredentialInfo,
+                              const void * pNetworkInterface )
+{
+    uint32_t ulPublishCount = 0U, ulTopicCount = 0U, ulDemoRunCount = 0U;
+    const uint32_t ulMaxPublishCount = 5UL;
+    NetworkContext_t xNetworkContext = { 0 };
+    MQTTContext_t xMQTTContext;
+    MQTTStatus_t xMQTTStatus;
+    TransportSocketStatus_t xNetworkStatus;
+    BaseType_t xTimerStatus;
+    BaseType_t xDemoStatus = pdPASS;
+    BaseType_t xIsConnectionEstablished = pdFALSE;
+
+    /* Remove compiler warnings about unused parameters. */
+    ( void ) awsIotMqttMode;
+    ( void ) pIdentifier;
+    ( void ) pNetworkServerInfo;
+    ( void ) pNetworkCredentialInfo;
+    ( void ) pNetworkInterface;
+
+
+    ulGlobalEntryTimeMs = prvGetTimeMs();
+
+    for( ; ulDemoRunCount < democonfigMQTT_MAX_DEMO_COUNT; ulDemoRunCount++ )
+    {
+        /* Serialize a PINGREQ packet to send upon invoking the keep-alive timer
+         * callback. */
+        xMQTTStatus = MQTT_SerializePingreq( &xPingReqBuffer );
```

Comment:
Can this be kept before the loop? The buffer needs to be populated only once with the PINGREQ packet

Reply:
Then I would have to put the loop into an if case.

Reply:
I was having a brain fart on some logic, moved it out.